llvm/lib/Target/AMDGPU/AMDGPUConvertWaveSize.cpp

```diff
@@ -0,0 +1,403 @@
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+#include "llvm/InitializePasses.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "si-convert-wave-size"
+
+namespace {
+class AMDGPUConvertWaveSize {
+  const GCNTargetMachine *TM;
+  const LoopInfo *LI;
+  ScalarEvolution *SE;
+  TargetTransformInfo *TTI;
+
+  InstructionCost TotalCost = 0;
+
+  static const unsigned MaxLatency = 2000;
+
+  SmallVector<Function *> Callees;
+
+public:
+  AMDGPUConvertWaveSize(const GCNTargetMachine *TM, const LoopInfo *LI,
+                        ScalarEvolution *SE, TargetTransformInfo *TTI)
+      : TM(TM), LI(LI), SE(SE), TTI(TTI) {}
+
+  bool run(Function &F);
+
+  bool changeWaveSizeAttr(Function *F);
+};
+
+class AMDGPUConvertWaveSizeLegacy : public FunctionPass {
+  const GCNTargetMachine *TM;
+
+public:
+  static char ID;
+  AMDGPUConvertWaveSizeLegacy(const GCNTargetMachine *TM) : FunctionPass(ID), TM(TM) {}
+  bool runOnFunction(Function &F) override {
+    auto &LI = getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+    auto &SE = getAnalysis<ScalarEvolutionWrapperPass>().getSE();
+    auto &TTI = getAnalysis<TargetTransformInfoWrapperPass>().getTTI(F);
+    AMDGPUConvertWaveSize Impl(TM, &LI, &SE, &TTI);
+    return Impl.run(F);
+  }
+  StringRef getPassName() const override { return "AMDGPU convert wave size"; }
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<LoopInfoWrapperPass>();
+    AU.addRequired<ScalarEvolutionWrapperPass>();
+    AU.setPreservesAll();
+    FunctionPass::getAnalysisUsage(AU);
+  }
+};
+} // end anonymous namespace
+
+void printFunctionAttributes(const Function &F) {
+  LLVM_DEBUG(dbgs() << "Function: " << F.getName() << "\n");
+  for (const auto &Attr : F.getAttributes()) {
+    LLVM_DEBUG(dbgs() << "  Attribute: " << Attr.getAsString() << "\n");
+  }
+}
+
+bool AMDGPUConvertWaveSize::run(Function &F) {
+
+  // Check if the function is a kernel.
+  if (F.getCallingConv() != CallingConv::AMDGPU_KERNEL)
+    return false;
+
+  const GCNSubtarget &ST = TM->getSubtarget<GCNSubtarget>(F);
+  if (!ST.isWave32()) {
+    LLVM_DEBUG(dbgs() << "AMDGPUConvertWaveSize: Kernel is not wave32.\n");
+    return false;
+  }
+
+  for (const auto &Arg : F.args()) {
+    if (Arg.getType()->isPointerTy() &&
+        Arg.getType()->getPointerAddressSpace() == AMDGPUAS::LOCAL_ADDRESS) {
+      LLVM_DEBUG(dbgs() << "AMDGPUConvertWaveSize: Kernel argument " << Arg
+                        << " points to LDS object\n");
+      return false;
+    }
+  }
+
+  // Check if the function can be called via device enqueue.
+  bool addressEscapes = false;
+  if (!F.use_empty()) {
+    const Module *M = F.getParent();
+    for (const GlobalVariable &GV : M->globals()) {
+      if (GV.hasInitializer()) {
+        if (const Constant *Init = GV.getInitializer()) {
+          if (isa<Function>(Init) && Init == &F) {
+            addressEscapes = true;
+          }
+        }
+      }
+    }
+  }
+
+  if (addressEscapes) {
```

Comment:
F.hasAddressTaken (but again, this whole thing would be easier to implement in attributor)

Reply:
It would be overkill, since this pass gives up if there is any function call so there is no inter-procedural analysis required.